consensus/raft/raft.go

```diff
@@ -686,3 +695,47 @@ func find(s []string, elem string) bool {
 	}
 	return false
 }
+
+func (rw *raftWrapper) observePeers() {
+	obsCh := make(chan hraft.Observation, 1)
+	defer close(obsCh)
+
+	if !sixtyfour {
+		// 32-bit systems don't support observers
+		// TODO: Check if this is still the case?
+		logger.Warning("raft Observers panic on 32-bit architecture")
+		return
+	}
+
+	observer := hraft.NewObserver(obsCh, false, nil)
+	rw.raft.RegisterObserver(observer)
+	defer rw.raft.DeregisterObserver(observer)
+
+	for {
+		select {
+		case obs := <-obsCh:
+			switch obs.Data.(type) {
+			case hraft.PeerObservation:
+				pObs := obs.Data.(hraft.PeerObservation)
+				if pObs.Removed {
+					logger.Infof("Observed removal of raft peer: %s",
+						pObs.Peer.ID)
+					pID, err := peer.IDB58Decode(string(pObs.Peer.ID))
+					if err != nil {
+						logger.Error(err)
+						continue
+					}
+					err = rw.peerManager.RmPeer(pID)
+					if err != nil {
+						logger.Error(err)
+					}
+				}
+			default:
+			}
+		case <-rw.finishObs:
+			close(rw.finishObs)
```

Comment:
channels should be closed by the senders, never by the receivers. But you want to use a context anyways. Have the raftWrapper keep a context that can be cancelled on shutdown.